src/apps/vguest/guest_vring.lua

```diff
@@ -0,0 +1,139 @@
+
+
+local ffi    = require("ffi")
+local C      = ffi.C
+local memory = require('core.memory')
+local packet = require('core.packet')
+local band = require('bit').band
+
+local VRing = {}
+VRing.__index = VRing
+
+local ringtypes = {}
+local function vring_type(n)
+   if ringtypes[n] then return ringtypes[n] end
+
+   local rng = ffi.typeof([[
+      struct {
+         struct {
+            uint64_t addr;
+            uint32_t len;
+            uint16_t flags;
+            uint16_t next;
+         } desc[$]          __attribute__((aligned(8)));
+         struct {
+            uint16_t flags;
+            uint16_t idx;
+            uint16_t ring[$];
+         } avail            __attribute__((aligned(8)));
+         struct {
+            uint16_t flags;
+            uint16_t idx;
+            struct {
+               uint32_t id;
+               uint32_t len;
+            } ring[$];
+         } used             __attribute__((aligned(4096)));
+      }
+   ]], n, n, n)
+   local t = ffi.typeof([[
+      struct {
+         int num, num_free;
+         uint16_t free_head, last_used_idx;
+         $ *vring;
+         uint64_t vring_physaddr;
+         struct packet *packets[$];
+      }
+   ]], rng, n)
+   ffi.metatype(t, VRing)
+   ringtypes[n] = t
+   return t
+end
+
+
+-- local function allocate_vring(n)
+--    local ct = vring_type(n)
+--    local ptr, phys, sz = memory.dma_alloc(ffi.sizeof(ct))
+--    ffi.fill(ptr, ffi.sizeof(ct))
+--    ptr = ffi.cast(ffi.typeof('$ *', ct), ptr)
+--    local obj = ptr[0]
+--
+--    -- arrange descs in a free list
+--    for i = 0, n-1 do
+--       obj.vring.desc[i].next = i+1
+--    end
+--    obj.num_free = n;
+--
+--    return obj, phys, sz
+-- end
+
```

Comment:
Remove commented code. There's more commented code in this file and other files too.